pkg/apis/batch/v1alpha1/job.go

```diff
@@ -225,29 +225,50 @@ const (
 	Inqueue JobPhase = "Inqueue"
 )
 
-// JobState contains details for the current state of the job.
-type JobState struct {
-	// The phase of Job.
-	// +optional
-	Phase JobPhase `json:"phase,omitempty" protobuf:"bytes,1,opt,name=phase"`
-
-	// Unique, one-word, CamelCase reason for the phase's last transition.
-	// +optional
-	Reason string `json:"reason,omitempty" protobuf:"bytes,2,opt,name=reason"`
+type JobConditionType string
 
-	// Human-readable message indicating details about last transition.
-	// +optional
-	Message string `json:"message,omitempty" protobuf:"bytes,3,opt,name=message"`
+const (
+	// JobCreated means the job has been accepted by the system.
+	JobCreated JobConditionType = "JobCreated"
+	// JobScheduled means the job has been successfully scheduled and at least
+	// 'MinAvailable' pods are running.
+	JobScheduled JobConditionType = "JobScheduled"
+	// JobSucceed means the job has been successfully executed.
+	JobSucceed JobConditionType = "JobSucceed"
+	// JobStopped means the job stopped without successfully execution.
+	// There could be several reasons for this condition
+	// 1. the system failed to executed the job after a max retry.
+	// 2. this job has been aborted or terminated.
+	JobStopped JobConditionType = "JobStopped"
+	// JobRestarting means the job is being restarting now. When in Restarting
+	// Condition the JobScheduled/JobSucceed/JobStopped conditions would be removed.
+	JobRestarting JobConditionType = "JobRestarting"
+)
 
-	// Last time the condition transit from one phase to another.
-	// +optional
-	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+// +k8s:deepcopy-gen=true
+// JobCondition describes the state of the job at a certain point.
+type JobCondition struct {
+	// Type of job condition.
+	Type JobConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+	// The reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+	// A human readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
 }
 
 // JobStatus represents the current status of a Job
 type JobStatus struct {
 	// Current state of Job.
-	State JobState `json:"state,omitempty" protobuf:"bytes,1,opt,name=state"`
+	Phase JobPhase `json:"phase,omitempty" protobuf:"bytes,1,opt,name=phase"`
```

Comment:
We lost the reason/message/lastTransitionTime field in `job.status.state` right? ... 

So, how does we prepare to maintain this information now?

Reply:
I don't found any usage for these two attributes(reason/message). if we need some additional informations, we can leverage the Condition's message/message or events. and for lastTransitionTime, it has been upgraded into Condition's lastTransitionTime attribute.

Reply:
seems good  :)